### ISSUE TITLE
Add DABL support to the DAML Assistant

### ISSUE THREAD

**Issue by bame-da** [digital-asset/daml#2829]
This issue is to discuss integrating DABL into DAML SDK functionality by adding assistant functionality for managing DABL projects. Main questions to answer:

Should we do it?
What should be possible through the CLI?
What's the UX?
Who maintains and tests it?

My initial thoughts:

Should we do it?
----------------

Yes, we should present DABL as the fastest route to production for simple use-cases that do not require to be distributed or consume the Ledger API. In these cases it makes sense to start developing against DABL from the get-go so it should be as easy to use as Sandbox/Navigator from our SDK.

What should be possible through the CLI?
------------------------------------------

At least the same things as for other ledgers: Uploading packages and administering parties.
I would restrict to that in the first version and postpone additional/special functionality like ledger reset to later iterations

What's the UX?
---------------

I'm pro-uniformity. Ideal UX in my mind would be to specify the DABL ledger in `daml.yaml` and make `daml deploy` and `daml ledger` DABL-aware.

E.g. specify the DABL ledger with

```
- ledger:
  - dabl: 987zyx
  - jwt: ABC123
```

Where `987zyx` is the DABL Ledger id, which translates directly into API endpoints. Then use `daml deploy` and `daml ledger` as usual, with different API calls being performed in the background. `daml ledger navigator` could show a warning that Navigator isn't supported, and start a browser with the DABL UI (e..g https://console.projectdabl.com/ledger/987zyx/active-contracts)

Maintenance and Testing
-------------------------

We can't easily test this from the DAML repo so testing would be on team-dabl. As long as the DABL admin API doesn't break all the time, the language team can support maintenance, but it's up to team-dabl to flag up when changes are needed.

@esaraniecki-da @da-tanabe, could you add any available documentation of your admin API here, please?
@moritzkiefer-da @fmota-da @nickchapman-da , once we have those docs, could you estimate effort to extend the assistant in this direction?

**Comment by cocreature** [comment]
Putting the jwt token in `daml.yaml` seems a bit problematic since we expect people to check in `daml.yaml` into the repository so at least for open source projects that is not an option and even for private projects it’s rather questionable.

**Comment by bame-da** [comment]
Good point. Maybe a pointer to a file containing the token, then, or read it from an environment variable?

**Comment by cocreature** [comment]
Both sound like reasonable options. Happy to let someone else make the decision here. One thing worth keeping in mind that we might not want to make this DABL specific as we add authentication to more parts of the stack.

**Comment by bame-da** [comment]
Yes, fair point, we'll need a good way for the assistant to pick up Admin JWTs for all kinds of ledgers. @gerolf-da , do you have opinions on what mechanism would make sense here?

Reading the admin JWT from a file with a default name (e.g. `.jwt`) might make it quite easy to keep the UX for other assistant functions smooth. `daml start` and `daml sandbox` could write that file by default, and other tools could pick it up by default, meaning if you don't do anything special, the master key to everything is quietly passed around via this file in the background.

**Comment by bame-da** [comment]
@esaraniecki-da @da-tanabe, with gRPC authentication in the pipeline, an alternative approach would be to expose the gRPC API and implement your own version of the package and party management services. That way DAML would behave like other DAML ledgers out of the box.

**Comment by gerolf-da** [comment]
> Reading the admin JWT from a file with a default name (e.g. `.jwt`) might make it quite easy to keep the UX for other assistant functions smooth. `daml start` and `daml sandbox` could write that file by default, and other tools could pick it up by default, meaning if you don't do anything special, the master key to everything is quietly passed around via this file in the background.

I don't have a clear picture in my mind yet. My mental model for sandbox was to turn api authentication off by default, as it doesn't really add anything to the development workflow other than a barrier, because users suddenly need to manage tokens.

I'm not sure I'm a fan of automatically writing a token file. What should happen if I run `daml sandbox` twice with a different `ledgerId`?

**Comment by da-tanabe** [comment]
I'm not sure how I managed to miss this issue (or maybe I just forgot about it)…

Now that auth seems largely in place, I think the only relevant configuration parameter that is specific to DABL is the ledger ID. However even this could potentially be expressed as a hostname (as discussed with @bame-da ): `api-<ledger-id>.projectdabl.com` or some other convention like this.

Would this lead to a consistent configuration mechanism across the board?

**Comment by bame-da** [comment]
Yes, pretty much. We won't have standardised IAM for all ledgers so you'll need a nice way to get hold of the DABL admin JWT somehow. Given you have a mechanism for that, once you expose gRPC, the normal `daml ledger` assistant should work for DABL.

**Comment by hurryabit** [comment]
@da-tanabe Do you expose gRPC by now?

**Comment by cocreature** [comment]
gRPC cannot be exposed in a usable form (TLS is garbage). I think DABL still exposes the package uploading endpoint but i don’t think it will ever be usable. I think the right solution here is to fix https://github.com/digital-asset/daml/issues/4824 and then support this on the DABL side (I believe at the moment they don’t expose the admin endpoints on the JSON API).
